co_sim_io/python/model_part_to_python.hpp

```diff
@@ -0,0 +1,79 @@
+//     ______     _____ _           ________
+//    / ____/___ / ___/(_)___ ___  /  _/ __ |
+//   / /   / __ \\__ \/ / __ `__ \ / // / / /
+//  / /___/ /_/ /__/ / / / / / / // // /_/ /
+//  \____/\____/____/_/_/ /_/ /_/___/\____/
+//  Kratos CoSimulationApplication
+//
+//  License:         BSD License, see license.txt
+//
+//  Main authors:    Philipp Bucher (https://github.com/philbucher)
+//
+
+#ifndef CO_SIM_IO_MODEL_PART_TO_PYHON_INCLUDED
+#define CO_SIM_IO_MODEL_PART_TO_PYHON_INCLUDED
+
+// Exposure of the CoSimIO to Python
+
+// System includes
+#include <string>
+#include <sstream>
+
+// pybind includes
+#include <pybind11/pybind11.h>
+#include <pybind11/stl.h>
+
+// CoSimIO include
+#include "../impl/model_part.hpp"
+
+void AddCoSimIOModelPartToPython(pybind11::module& m)
+{
+    namespace py = pybind11;
+
+    py::class_<CoSimIO::Node, CoSimIO::ModelPart::NodePointerType>(m,"Node")
+        .def(py::init<const CoSimIO::IdType, const double, const double, const double>())
+        .def(py::init<const CoSimIO::IdType, const CoSimIO::CoordinatesType&>())
+        .def("Id", &CoSimIO::Node::Id)
+        .def("X",  &CoSimIO::Node::X)
+        .def("Y",  &CoSimIO::Node::Y)
+        .def("Z",  &CoSimIO::Node::Z)
+        .def("Coordinates",  &CoSimIO::Node::Coordinates)
+        .def("__str__",   [](const CoSimIO::Node& I_Node)
+            { std::stringstream ss; ss << I_Node; return ss.str(); } )
+        ;
+
+    py::class_<CoSimIO::Element, CoSimIO::ModelPart::ElementPointerType>(m,"Element")
+        .def(py::init<const CoSimIO::IdType, const CoSimIO::Element::ElementType, const CoSimIO::Element::NodesContainerType&>())
+        .def("Id", &CoSimIO::Element::Id)
+        .def("Type", &CoSimIO::Element::Type)
+        .def("NumberOfNodes", &CoSimIO::Element::NumberOfNodes)
+        .def_property_readonly("Nodes", [](CoSimIO::Element& I_Element) {
+            return py::make_iterator(I_Element.NodesBegin(), I_Element.NodesEnd());
+            }, py::keep_alive<0, 1>()) /* Keep vector alive while iterator is used */
+        .def("__str__",   [](const CoSimIO::Element& I_Element)
+            { std::stringstream ss; ss << I_Element; return ss.str(); } )
+        ;
+
+    py::class_<CoSimIO::ModelPart>(m,"ModelPart")
```

Comment:
Note to self, maybe this needs to be updated to also have a holder type for passing to C++
=> ATM I think its not necessary as it is being passed by ref

Reply:
hmmmm, just to be clear, So you are assuming that the python would remove the modelpart after cleaning everything else in C++ side.

Reply:
I just wrote it other way around, the c++ will be the holder and python would be the one who should cleanup related things before Modelpart removed by c++

Reply:
I am confused :)
can we briefly speak next week? I think it would be good to fully understand this also for other things

Reply:
for the record, after speaking we agreed that `Node` and `Element` are held by shared ptrs bcs they can be created in python and passed to the `ModelPart`
The python `ModelPart` however is held by python only and given to C++ by reference